### DISCUSSION TITLE
How to get custom font in a pdf (craft module setup)

### DISCUSSION THREAD

**Discussion by JulieVelghe** [dompdf/dompdf#3473] [Q&A]
Hi all,
I have a craft website project. Some entries feature a print button that should print out information about the entry in a pdf with custom styling. I'm using a module to generate the pdf. The loadHtml is located in the service file, that gets called by the controller to send back the pdf file.
![image](https://github.com/dompdf/dompdf/assets/46524074/7dcfa2a2-a21e-4702-8c2b-b1571cea434a)

The font I want to use is Mona sans which is located in my www folder:
![image](https://github.com/dompdf/dompdf/assets/46524074/3220ba08-1876-440d-9974-da707bec6805)

How to I make sure that I can use that font? I'm using a twig template file where I define my font faces like this:
```
@font-face {
    font-family: 'Mona Sans Wide';
    font-style: normal;
    font-weight: 700;
    src: url({{ "#{currentSite.baseUrl}assets/dist/fonts/Mona-Sans/Mona-Sans-BoldWide.ttf" }}) format('truetype');
}
```
But this doesn't work.

Any help is very much appreciated.

**Comment by bsweeney** [comment]
It depends on what currentSite.baseUrl provides. Is that a filesystem path or a web URL. Either way, you'll need to make sure you've configured the appropriate access permissions.

For a web URL you just need to enable remote file access, for example:

```
$dompdf = new Dompdf(["isRemoteEnabled" => true]);
```

For a filesystem path you would need to add that path to the allowable paths, for example:

```
$dompdf = new Dompdf(["chroot" => [$currentSite.baseUrl]]);
```

(guessing that variable reference is maybe different when called within the service file, but you get the picture)

**Reply by JulieVelghe** [reply]
that variable puts out 'https://local.referenties.karakters.be'. 
I have enabled the remote file access like this:
        $options->setIsRemoteEnabled(true);
        $dompdf->setOptions($options);
But that doesn't seem to work. Also, changing the font-face to use the url instead of the variable doesn't work out.

**Reply by bsweeney** [reply]
Do you need a "/" in front of assets then?

It's also possible that PHP can't resolve the URL. You could confirm this with a small bit of logic (e.g., `file_get_contents("https://local.referenties.karakters.be/assets/dist/fonts/Mona-Sans/Mona-Sans-BoldWide.ttf")`).

If you want to dive in a bit more, you can try enabling some of the [debug options](https://github.com/dompdf/dompdf/wiki/Troubleshooting-Dompdf).

**Reply by JulieVelghe** [reply]
Thank you @bsweeney for the help already.
I've tried that file_get_contents on the font file, and I'm getting errors:
PHP Warning – [yii\base\ErrorException](https://www.yiiframework.com/doc-2.0/yii-base-errorexception.html)
file_get_contents(): SSL operation failed with code 1. OpenSSL Error messages:
error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed
Is this something that I can solve in htaccess perhaps?

**Reply by bsweeney** [reply]
If you can do it over HTTP that would probably work OK. Otherwise maybe try disabling verification of the SSL cert using a [custom HTTP context](https://github.com/dompdf/dompdf/wiki/Common-Use-Cases#custom-http-stream-context). Maybe something like this:

```
$options->setHttpContext([
    'ssl' => [
        'verify_peer' => false,
        'verify_peer_name' => false,
        'allow_self_signed' => true
    ]
];
```

**Reply by JulieVelghe** [reply]
Yes, that custom http context seemed to do the trick now. Thank you so much for your help!
